proxy/http2/Http2ConnectionState.cc

```diff
@@ -1601,7 +1595,22 @@ Http2ConnectionState::send_goaway_frame(Http2StreamId id, Http2ErrorCode ec)
   SCOPED_MUTEX_LOCK(lock, this->ua_session->mutex, this_ethread());
   this->ua_session->handleEvent(HTTP2_SESSION_EVENT_XMIT, &frame);
 
-  handleEvent(HTTP2_SESSION_EVENT_FINI, nullptr);
+  if (!this->ua_session->get_half_close_flag()) {
+    this->ua_session->set_half_close_flag(true);
+    DebugHttp2Con(ua_session, "session half-close");
+  }
+
+  // As for gracefull shutdown, TS should process outstanding stream as long as possible.
+  // As for signal connection error, TS should close connection immediately.
+  if (ec != Http2ErrorCode::HTTP2_ERROR_NO_ERROR) {
+    HTTP2_INCREMENT_THREAD_DYN_STAT(HTTP2_STAT_CONNECTION_ERRORS_COUNT, this_ethread());
+
+    UnixNetVConnection *client_vc = static_cast<UnixNetVConnection *>(this->ua_session->get_netvc());
+    if (client_vc) {
+      client_vc->add_to_active_queue();
+      client_vc->set_active_timeout(HRTIME_SECONDS(1));
```

Comment:
From this approach, I got a crash with h2spec. I don't see this crash on master branch. I need to dig more.

```
h2spec http2/6.1/3 -p 4443 -k -t -v
```

```
(lldb) bt
* thread #12, name = '[ET_NET 9]', stop reason = signal SIGABRT
  * frame #0: 0x00007fffc096dd42 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fffc0a5b5bf libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fffc08d3420 libsystem_c.dylib`abort + 129
    frame #3: 0x000000000002a3c0 libtsutil.7.dylib`ink_abort(message_format="%s:%d: failed assertion `%s`") at ink_error.cc:99
    frame #4: 0x00000000000271ff libtsutil.7.dylib`::_ink_assert(expression="c_write > 0", file="HttpTunnel.cc", line=981)at ink_assert.cc:37
    frame #5: 0x00000001001a5ff1 traffic_server`HttpTunnel::producer_run(this=0x00000000068f9cf8, p=0x00000000068f9f00) atHttpTunnel.cc:981
    frame #6: 0x00000001001a564d traffic_server`HttpTunnel::tunnel_run(this=0x00000000068f9cf8, p_arg=0x00000000068f9f00) at HttpTunnel.cc:797
    frame #7: 0x0000000100147f55 traffic_server`HttpSM::do_setup_post_tunnel(this=0x00000000068f89b0, to_vc_type=HTTP_SERVER_VC) at HttpSM.cc:5756
    frame #8: 0x000000010014732a traffic_server`HttpSM::state_send_server_request_header(this=0x00000000068f89b0, event=103, data=0x0000000001124ea0) at HttpSM.cc:2021
    frame #9: 0x00000001001374e8 traffic_server`HttpSM::main_handler(this=0x00000000068f89b0, event=103, data=0x0000000001124ea0) at HttpSM.cc:2682
    frame #10: 0x000000010001f200 traffic_server`Continuation::handleEvent(this=0x00000000068f89b0, event=103, data=0x0000000001124ea0) at I_Continuation.h:153
    frame #11: 0x000000010035dfb8 traffic_server`write_signal_and_update(event=103, vc=0x0000000001124d00) at UnixNetVConnection.cc:176
    frame #12: 0x000000010035e1cd traffic_server`write_signal_done(event=103, nh=0x000000000380be80, vc=0x0000000001124d00) at UnixNetVConnection.cc:218
    frame #13: 0x000000010035db0f traffic_server`write_to_net_io(nh=0x000000000380be80, vc=0x0000000001124d00, thread=0x0000000003808000) at UnixNetVConnection.cc:609
    frame #14: 0x000000010035d0a5 traffic_server`write_to_net(nh=0x000000000380be80, vc=0x0000000001124d00, thread=0x0000000003808000) at UnixNetVConnection.cc:443
    frame #15: 0x000000010035117b traffic_server`NetHandler::mainNetEvent(this=0x000000000380be80, event=5, e=0x000000000182e6a0) at UnixNet.cc:526
    frame #16: 0x000000010001f200 traffic_server`Continuation::handleEvent(this=0x000000000380be80, event=5, data=0x000000000182e6a0) at I_Continuation.h:153
    frame #17: 0x00000001003827d5 traffic_server`EThread::process_event(this=0x0000000003808000, e=0x000000000182e6a0, calling_code=5) at UnixEThread.cc:143
    frame #18: 0x00000001003830cd traffic_server`EThread::execute(this=0x0000000003808000) at UnixEThread.cc:270
    frame #19: 0x0000000100381d17 traffic_server`spawn_thread_internal(a=0x00000000007002c0) at Thread.cc:84
    frame #20: 0x00007fffc0a589af libsystem_pthread.dylib`_pthread_body + 180
    frame #21: 0x00007fffc0a588fb libsystem_pthread.dylib`_pthread_start + 286
    frame #22: 0x00007fffc0a58101 libsystem_pthread.dylib`thread_start + 13
```